### PR TITLE
feature: NoTrailingCommaInSinglelineFixer - Introduction

### PR DESCRIPTION
**Pull request by SpacePossum**
Currently there is no fixer to enforce no trailing `,` in a single of short array destruction in one line, nor  a rule for no trailing `,` in single line of group imports.
Other fixes are; it is now single rule with options and all fixing is done in same matter (no longer a priority case).

### CODE REVIEW

File: tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+            '<?php $b = foo($a,);',
+        ];
+
+        yield '.' => [
+            '<?php $c = $b . foo($a);',
+            '<?php $c = $b . foo($a,);',
+        ];
+
+        yield '(' => [
+            '<?php (foo($a/* 1X */   /* 2 */  ));',
```
Comment (julienfalque) [inline tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php:90]: IMO this should be:
```suggestion
            '<?php (foo($a /* 1X */ /* 2 */  ));',
```
i.e. removing the spaces that are directly before the comma, not directly after the previous element.

File: tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    public function testFix80NoTrailingCommaInSinglelineFunctionCallFixer(string $expected, string $input = null): void
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public static function provideFix80NoTrailingCommaInSinglelineFunctionCallFixerCases(): iterable
+    {
+        yield [
+            '<?php function foo(
+    #[MyAttr(1, 2,)] Type $myParam,
```
Comment (julienfalque) [inline tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php:240]: Why not removing the trailing comma in the attribute?

File: tests/Fixer/FunctionNotation/NoTrailingCommaInSinglelineFunctionCallFixerTest.php [hunk 3]
```diff
@@ -59,7 +59,7 @@ public function provideFixCases(): iterable
         ];
 
         yield '(' => [
-            '<?php (foo($a /* 1 */   /* 2 */  ));',
+            '<?php (foo($a/* 1 */   /* 2 */  ));',
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixer/FunctionNotation/NoTrailingCommaInSinglelineFunctionCallFixerTest.php:62]: The fixer now consistently fixes in one way, the current different fixers have so little differences in handling edges cases like these.

File: tests/AutoReview/ProjectCodeTest.php [hunk 4]
```diff
@@ -319,8 +319,6 @@ public function testThatTestClassCoversAreCorrect(string $testClassName): void
         static::assertNotFalse($doc);
 
         if (1 === Preg::match('/@coversNothing/', $doc, $matches)) {
-            $this->expectNotToPerformAssertions();
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/AutoReview/ProjectCodeTest.php:322]: On line 319 an assertion is done `static::assertNotFalse($doc);`, as such PHPUnit warns about the false statement not to expect assertions.

File: tests/Fixer/ArrayNotation/NoTrailingCommaInSinglelineArrayFixerTest.php [hunk 5]
```diff
@@ -50,11 +50,6 @@ public function provideFixCases(): array
                 '<?php
     $test = array("foo", <<<TWIG
         foo
-TWIG
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixer/ArrayNotation/NoTrailingCommaInSinglelineArrayFixerTest.php:53]: The fixer is no consistent.
This case:
```php
<?php
    $test = array("foo", <<<TWIG
        foo
TWIG
        , $twig, );
```
is a multi line array, as such removing the last comma is incorrect.
The current implementation considers a multi line heredoc/nowdoc as a single line, this bug is now fixed.

File: src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        $elements = ['arguments', 'array_destructuring', 'array', 'group_import'];
+
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('elements', 'Which elements to fix.'))
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php:65]: It was mainly done for the deprecated fixers indeed. Maybe we can consider removing these next major if there really is no interest in these?

File: tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php (hunk 1 above)
Comment (SpacePossum) [reply tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php:90]: Some fixers did one thing, others the other thing, it is a real edge case, I'm not sure it is worth the logic to handle this differently.

File: tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php (hunk 2 above)
Comment (SpacePossum) [reply tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php:240]: This is a test that is not touched:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v3.10.0/tests/Fixer/FunctionNotation/NoTrailingCommaInSinglelineFunctionCallFixerTest.php#L214

Comment (SpacePossum) [reply tests/Fixer/Basic/NoTrailingCommaInSinglelineFixerTest.php:240]: Would be nice to add support for it, but it is not the goal of the PR and it is already pretty big.

File: src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+final class NoTrailingCommaInSinglelineFixer extends AbstractFixer implements ConfigurableFixerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'In a single line statement there MUST NOT be a trailing comma after: the last argument when calling a function, the last element when declaring array, the last variable assignment when destructuring, the last `use` group when declaring usage.',
```
Comment (jrmajor) [inline src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php:38]:
```suggestion
            'If a list of values separated by a comma is contained on a single line, then the last item MUST NOT have a trailing comma.',
```

Comment (SpacePossum) [reply src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php:38]: Thanks for the suggestion, if others could :+1:  or :-1:  my comment here than I know if we want this new description proposal or if we want to go with the description as proposed currently in the PR, thanks!

File: src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (julienfalque) [reply src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php:65]: If we start with a public option, we will likely forget to consider removing it on next major release. Using e.g. a constructor argument will allow us to detect when it's not used anymore and drop it.

[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php:65]: I'm not sure I follow, the construct also has to be public, why wouldn't forget one and not the other?

[the remark this answers is not in the document]
Comment (julienfalque) [reply src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php:65]: Ah my bad, I though the fixers were `@internal`.

File: src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php (hunk 7 above)
Comment (SpacePossum) [reply src/Fixer/Basic/NoTrailingCommaInSinglelineFixer.php]: docs are updated

**Review by julienfalque — APPROVED** [review]

**Comment by julienfalque** [comment]
Thank you @SpacePossum.

**Comment by SpacePossum** [comment]
Thanks all reviewers and @julienfalque for merging :D  :dancers:

**Pull request merged**